### PR TITLE
Add two events Pty/Set_env/Start_shell into the server

### PR DESCRIPTION
**Pull request by dinosaure**
This PR change a bit the interface of `awa-mirage` and give the ability to handle new events:
- `Pty`: when the window change
- `Set_env`: when the client wants to set the local environment
- `Shell`: when the client requests a shell

### CODE REVIEW

File: lwt/awa_lwt.ml [hunk 1]
```diff
@@ -157,6 +157,7 @@ let rec nexus t fd server input_buffer =
       let c = { cmd; id; sshin_mbox; exec_thread } in
       let t = { t with channels = c :: t.channels } in
       nexus t fd server input_buffer
+    | _ -> nexus t fd server input_buffer
```
Comment (hannesm) [inline lwt/awa_lwt.ml:160]: can we avoid a catch-all here, please? :)

File: mirage/awa_mirage.ml [hunk 2]
```diff
@@ -96,8 +96,7 @@ module Make (F : Mirage_flow.S) (T : Mirage_time.S) (M : Mirage_clock.MCLOCK) =
           | `Nothing, `Channel_eof _ -> `Eof
           | `Nothing, `Disconnected -> `Eof
           | a, `Channel_stderr (id, data) ->
-            Log.warn (fun m -> m "%ld stderr %s" id (Cstruct.to_string data));
-            a
+            Log.warn (fun m -> m "%ld stderr %s" id (Cstruct.to_string data)); a
```
Comment (hannesm) [inline mirage/awa_mirage.ml:99]: I find the previous (putting the log and `a` on different lines) more readable.

**Comment by hannesm** [comment]
LGTM, minor nits in the comments.

**Comment by dinosaure** [comment]
I decided to upgrade the `lwt` layer with the support of the shell in a different way than the `mirage` layer via a _callback_ for `set_env` and `set_window` (when we got such event from the server) and an optional `?cmd` argument for the `exec_callback`. It breaks the API too but less than the `mirage` layer.

**Comment by dinosaure** [comment]
I think it's ready to merge. A question remains about an homogeneized interface between `awa-lwt` and `awa-mirage` but I would like to have a feedback on that before. /cc @hannesm @reynir (@palainp already updated `sshfs` accordind to this PR).

**Comment by palainp** [comment]
> (@palainp already updated `sshfs` accordind to this PR).

Unfortunately not yet, but I will :)

> A question remains about an homogeneized interface between `awa-lwt` and `awa-mirage` but I would like to have a feedback on that before. 

I prefer the awa-mirage style, but that's just a personal view, and I agree that a similar interface should be beneficial.

**Comment by hannesm** [comment]
I'm fine with this PR. I did not use the server API (yet).

**Pull request merged**